www/data_service/src/services/data/data.service.coffee

```diff
@@ -0,0 +1,184 @@
+# TODO rename this, or the dataService in waterfall view
+class Data extends Provider
+    constructor: ->
+    # TODO caching
+    cache: false
+
+    $get: ($log, $injector, $q, restService, socketService, dataUtilsService, ENDPOINTS) ->
+        return new class DataService
+            self = null
+            constructor: ->
+                self = @
+                # setup socket listeners
+                socketService.eventStream.onUnsubscribe = @unsubscribeListener
+                socketService.socket.onclose = @socketCloseListener
+                # generate loadXXX functions for root endpoints
+                @constructor.generateEndpoints()
+
+            # the arguments are in this order: endpoint, id, child, id of child, query
+            get: (args...) ->
+                # keep defined arguments only
+                args = args.filter (e) -> e?
+
+                # get the query parameters
+                [..., last] = args
+                # subscribe for changes if 'subscribe' is true or undefined
+                subscribe = last.subscribe or not last.subscribe?
+                if angular.isObject(last)
+                    query = args.pop()
+                    # 'subscribe' is not part of the query
+                    delete query.subscribe
+
+                # up to date array, this will be returned
+                updating = []
+
+                promise = $q (resolve, reject) =>
+
```

Comment:
I think there might be to many callbacks nested here. Is there a way to reduce it a little? Resolving of some promises such as the `socketPromise` and `restPromise` seems can be done in parallel. Maybe use `$q.all` to combine several promises is better.